components/service/glean/src/main/java/mozilla/components/service/glean/storages/StringListsStorageEngine.kt

```diff
@@ -0,0 +1,133 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.service.glean.storages
+
+import android.annotation.SuppressLint
+import android.content.SharedPreferences
+import mozilla.components.service.glean.CommonMetricData
+import mozilla.components.support.base.log.logger.Logger
+import org.json.JSONArray
+import java.lang.Exception
+
+/**
+ * This singleton handles the in-memory storage logic for string lists. It is meant to be used by
+ * the Specific String List API and the ping assembling objects. No validation on the stored data
+ * is performed at this point: validation must be performed by the Specific String List API.
+ *
+ * This class contains a reference to the Android application Context. While the IDE warns
+ * us that this could leak, the application context lives as long as the application and this
+ * object. For this reason, we should be safe to suppress the IDE warning.
+ */
+@SuppressLint("StaticFieldLeak")
+internal object StringListsStorageEngine : StringListsStorageEngineImplementation()
+
+internal open class StringListsStorageEngineImplementation(
+    override val logger: Logger = Logger("glean/StringsListsStorageEngine")
+) : GenericScalarStorageEngine<List<String>>() {
+
+    companion object {
+        // Maximum length of any list
+        const val MAX_LIST_LENGTH_VALUE = 20
+    }
+
+    override fun deserializeSingleMetric(value: Any?): List<String>? {
+        // Since SharedPreferences doesn't directly support storing of List<> types, we must use
+        // an intermediate JSONArray which can be deserialized and converted back to List<String>.
+        // Using JSONArray introduces a possible issue in that it's constructor will still properly
+        // convert a stringified JSONArray into an array of Strings regardless of whether the values
+        // have been properly quoted or not.  For example, [1,2,3] is as valid just like
+        // ["a","b","c"] is valid.
+        return (value as? String)?.let {
+            val jsonArray: JSONArray
+            try {
+                jsonArray = JSONArray(it)
```

Comment:
Are you sure this throws? The docs state it only yields a [weird array](https://developer.android.com/reference/org/json/JSONArray.html#JSONArray(java.util.Collection)). Let's get rid of the try/catch if not needed.

Reply:
I’m sure it throws because it threw while I was trying to resolve the test case for this.  Hence the specific exception.  I can add an explicit comment to describe situations that would cause it to throw for documentation purposes if that helps.

Reply:
That's interesting! How lovely are outdated docs? :) Can you expand on how you triggered the exception there?

If we are to keep the try/catch here, I'd refactor it to be more idiomatic (`try` can return values!):

```
val jsonArray: JSONArray? = try {
  JSONArray(it)
} catch (...) {
  null
}

return@let jsonArray?.toList()
```

Reply:
```
"/Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java" -ea -Didea.test.cyclic.buffer.size=1048576 -Didea.launcher.port=50449 "-Didea.launcher.bin.path=/Applications/Android Studio.app/Contents/bin" -Dfile.encoding=UTF-8 -classpath "/Applications/Android Studio.app/Contents/lib/idea_rt.jar:/Applications/Android Studio.app/Contents/plugins/junit/lib/junit-rt.jar:/Applications/Android Studio.app/Contents/plugins/junit/lib/junit5-rt.jar:/Users/tlong/Library/Android/sdk/platforms/android-28/data/res:/Users/tlong/Documents/android-components/components/service/glean/build/intermediates/javac/debug/compileDebugJavaWithJavac/classes:/Users/tlong/Documents/android-components/components/service/glean/build/tmp/kotlin-classes/debug:/Users/tlong/Documents/android-components/components/service/glean/build/generated/res/rs/debug:/Users/tlong/Documents/android-components/components/service/glean/build/generated/res/resValues/debug:/Users/tlong/Documents/android-components/components/service/glean/build/tmp/kotlin-classes/debugUnitTest:/Users/tlong/Documents/android-components/components/service/glean/build/intermediates/unit_test_config_directory/debugUnitTest/generateDebugUnitTestConfig/out:/Users/tlong/Documents/android-components/components/service/glean/build/intermediates/sourceFolderJavaResources/test/debug:/Users/tlong/.gradle/caches/modules-2/files-2.1/backport-util-concurrent/backport-util-concurrent/3.1/682f7ac17fed79e92f8e87d8455192b63376347b/backport-util-concurrent-3.1.jar:/Users/tlong/.gradle/caches/transforms-1/files-1.1/localbroadcastmanager-28.0.0.aar/2757de20d4f49c291b72a5aa99f645e4/jars/classes.jar:/Users/tlong/.gradle/caches/transforms-1/files-1.1/support-vector-drawable-28.0.0.aar/1a5089fded569e54b3062af0d7a26898/jars/classes.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/org.objenesis/objenesis/2.6/639033469776fd37c08358c6b92a4761feb2af4b/objenesis-2.6.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-stdlib/1.3.10/b178c1501609c6e4ee8be635513cb023a466457d/kotlin-stdlib-1.3.10.jar:/Users/tlong/.gradle/caches/transforms-1/files-1.1/interpolator-28.0.0.aar/f1af72ec706f868a731754fbe585a316/jars/classes.jar:/Users/tlong/.gradle/caches/transforms-1/files-1.1/support-core-utils-28.0.0.aar/bf5266b995274f61ad8dfb18fb69e7f4/jars/classes.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/org.robolectric/robolectric/4.0.2/69d49494c7b22ca4af18c0d2a248b76fd971d730/robolectric-4.0.2.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/org.apache.maven/maven-error-diagnostics/2.2.1/e81bb342d7d172f23d108dc8fa979a1facdcde8e/maven-error-diagnostics-2.2.1.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/org.ow2.asm/asm-tree/7.0/29bc62dcb85573af6e62e5b2d735ef65966c4180/asm-tree-7.0.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/org.robolectric/sandbox/4.0.2/51429643ffa87b3dffa8efcdc0a323d584bd659b/sandbox-4.0.2.jar:/Users/tlong/.gradle/caches/transforms-1/files-1.1/drawerlayout-28.0.0.aar/e19d3ac3a3e1b199a248146429ab39fa/jars/classes.jar:/Users/tlong/.gradle/caches/transforms-1/files-1.1/coordinatorlayout-28.0.0.aar/52fc59e8620df1c55c9c87eb8a514b7d/res:/Users/tlong/.gradle/caches/transforms-1/files-1.1/coordinatorlayout-28.0.0.aar/52fc59e8620df1c55c9c87eb8a514b7d/jars/classes.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/org.apache.maven.wagon/wagon-http-lightweight/1.0-beta-6/b3815078570c3b1f0667e123d59717c6b726c6c2/wagon-http-lightweight-1.0-beta-6.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/org.apache.maven/maven-project/2.2.1/8239e98c16f641d55a4ad0e0bab0aee3aff8933f/maven-project-2.2.1.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/org.apache.maven.wagon/wagon-file/1.0-beta-6/6c53633505460caf49d2660de1e24744d915afb9/wagon-file-1.0-beta-6.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/org.apache.maven.wagon/wagon-provider-api/1.0-beta-6/3f952e0282ae77ae59851d96bb18015e520b6208/wagon-provider-api-1.0-beta-6.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/org.ow2.asm/asm-commons/7.0/478006d07b7c561ae3a92ddc1829bca81ae0cdd1/asm-commons-7.0.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/com.squareup.okhttp3/mockwebserver/3.10.0/3bb03fab2fbdef497eff17a96ddf50f3e30fd1fe/mockwebserver-3.10.0.jar:/Users/tlong/.gradle/caches/transforms-1/files-1.1/swiperefreshlayout-28.0.0.aar/80646cba0c65dfae2ee1573f41e7ef0b/jars/classes.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/org.apache.ant/ant/1.8.0/7b456ca6b93900f96e58cc8371f03d90a9c1c8d1/ant-1.8.0.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/org.jetbrains/annotations/13.0/919f0dfe192fb4e063e7dacadee7f8bb9a2672a9/annotations-13.0.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/org.mockito/mockito-core/2.23.0/497ddb32fd5d01f9dbe99a2ec790aeb931dff1b1/mockito-core-2.23.0.jar:/Users/tlong/.gradle/caches/transforms-1/files-1.1/livedata-1.1.1.aar/f4aec5355c9c427896e33127591d13b1/jars/classes.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/org.ow2.asm/asm-analysis/7.0/4b310d20d6f1c6b7197a75f1b5d69f169bc8ac1f/asm-analysis-7.0.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlinx/kotlinx-coroutines-core-common/1.0.1/679991d1b388fd9392eeb3d63b10b3f4c284f389/kotlinx-coroutines-core-common-1.0.1.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/com.squareup.okio/okio/1.14.0/102d7be47241d781ef95f1581d414b0943053130/okio-1.14.0.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-stdlib-jdk7/1.3.10/4d147bf43060dc43d61b096e24da1e67dfe0c032/kotlin-stdlib-jdk7-1.3.10.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/net.bytebuddy/byte-buddy/1.9.0/8cb0d5baae526c9df46ae17693bbba302640538b/byte-buddy-1.9.0.jar:/Users/tlong/.gradle/caches/transforms-1/files-1.1/versionedparcelable-28.0.0.aar/b65ce8141075eb1d6960cedb136f28cc/jars/classes.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/org.robolectric/resources/4.0.2/60865d27446b2be462382e35dbd8a4fa9f591455/resources-4.0.2.jar:/Users/tlong/.gradle/caches/transforms-1/files-1.1/runtime-1.1.1.aar/0f01894ac3c59a2706f28d3c841636a5/jars/classes.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/org.robolectric/annotations/4.0.2/85fe30eba3d9500615ede75f8a692a8e90ebcbff/annotations-4.0.2.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/org.robolectric/utils/4.0.2/81089716c38250441358df5ec862b604b0009d1f/utils-4.0.2.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/com.google.guava/guava/20.0/89507701249388e1ed5ddcf8c41f4ce1be7831ef/guava-20.0.jar:/Users/tlong/.gradle/caches/transforms-1/files-1.1/loader-28.0.0.aar/5a37718d1320e91c654c18b5190bab10/jars/classes.jar:/Users/tlong/.gradle/caches/transforms-1/files-1.1/core-1.0.0.aar/8a2a176dbbd17f6b277371271ae26585/res:/Users/tlong/.gradle/caches/transforms-1/files-1.1/core-1.0.0.aar/8a2a176dbbd17f6b277371271ae26585/jars/classes.jar:/Users/tlong/.gradle/caches/transforms-1/files-1.1/viewpager-28.0.0.aar/736e9879aeb9b023714e5d6bc070054e/jars/classes.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/com.android.support/support-annotations/28.0.0/ed73f5337a002d1fd24339d5fb08c2c9d9ca60d8/support-annotations-28.0.0.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/org.robolectric/shadowapi/4.0.2/93ac412550c18147f765479adb01dc077bd78f2/shadowapi-4.0.2.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/com.almworks.sqlite4java/sqlite4java/0.282/745a7e2f35fdbe6336922e0d492c979dbbfa74fb/sqlite4java-0.282.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlinx/kotlinx-coroutines-android/1.0.1/92ab20a9ba1378b29ca88d5ecbd442e9bb4186d3/kotlinx-coroutines-android-1.0.1.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/org.apache.maven/maven-artifact-manager/2.2.1/ec355b913c34d37080810f98e3f51abecbe1572b/maven-artifact-manager-2.2.1.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/org.apache.maven/maven-settings/2.2.1/2236ffe71fa5f78ce42b0f5fc22c54ed45f14294/maven-settings-2.2.1.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlinx/kotlinx-coroutines-core/1.0.1/f33e8dab753f33d1bbb07cca664fd6f13d993d7e/kotlinx-coroutines-core-1.0.1.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/net.bytebuddy/byte-buddy-agent/1.9.0/37b5703b4a6290be3fffc63ae9c6bcaaee0ff856/byte-buddy-agent-1.9.0.jar:/Users/tlong/.gradle/caches/transforms-1/files-1.1/runtime-1.1.1.aar/9112706e9e0b756e4c7a30f374dc3383/jars/classes.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/org.codehaus.plexus/plexus-container-default/1.0-alpha-9-stable-1/94aea3010e250a334d9dab7f591114cd6c767458/plexus-container-default-1.0-alpha-9-stable-1.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/org.bouncycastle/bcprov-jdk15on/1.52/88a941faf9819d371e3174b5ed56a3f3f7d73269/bcprov-jdk15on-1.52.jar:/Users/tlong/.gradle/caches/transforms-1/files-1.1/customview-28.0.0.aar/12146ddaac9b71567a0aeb07e8396728/jars/classes.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/org.apache.maven/maven-plugin-registry/2.2.1/72a24b7775649af78f3986b5aa7eb354b9674cfd/maven-plugin-registry-2.2.1.jar:/Users/tlong/.gradle/caches/transforms-1/files-1.1/support-core-ui-28.0.0.aar/cfc3b972106a9fb98e9840a63d7098df/jars/classes.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/org.apache.maven/maven-ant-tasks/2.1.3/b09be554228d66d208e5fef5266844aacf443abc/maven-ant-tasks-2.1.3.jar:/Users/tlong/.gradle/caches/transforms-1/files-1.1/slidingpanelayout-28.0.0.aar/674a07b4e87127baa88d4755084594da/jars/classes.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/com.ibm.icu/icu4j/53.1/786d9055d4ca8c1aab4a7d4ac8283f973fd7e41f/icu4j-53.1.jar:/Users/tlong/.gradle/caches/transforms-1/files-1.1/viewmodel-1.1.1.aar/053db607d529b1952b46416565e38f7b/jars/classes.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/com.google.protobuf/protobuf-java/2.6.1/d9521f2aecb909835746b7a5facf612af5e890e8/protobuf-java-2.6.1.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/com.android.support/collections/28.0.0/c1bcdade4d3cc2836130424a3f3e4182c666a745/collections-28.0.0.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/org.apache.maven.wagon/wagon-http-shared/1.0-beta-6/ccd70d7e0d8c085e648a83f072da06ca9a53be94/wagon-http-shared-1.0-beta-6.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/org.robolectric/junit/4.0.2/145a91e70af0c310830725ef37b7b202ca1b990d/junit-4.0.2.jar:/Users/tlong/.gradle/caches/transforms-1/files-1.1/documentfile-28.0.0.aar/0521f2ad3da83c07d170a83db2919c75/jars/classes.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/org.apache.ant/ant-launcher/1.8.0/8b53ba16fa62fb1034da8f1de200ddc407c8381/ant-launcher-1.8.0.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/org.apache.maven/maven-profile/2.2.1/3950071587027e5086e9c395574a60650c432738/maven-profile-2.2.1.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/androidx.lifecycle/lifecycle-common/2.0.0/e070ffae07452331bc5684734fce6831d531785c/lifecycle-common-2.0.0.jar:/Users/tlong/.gradle/caches/transforms-1/files-1.1/cursoradapter-28.0.0.aar/b33d73847cda8937af7ba3090a7ffd4c/jars/classes.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/nekohtml/nekohtml/1.9.6.2/2d960be7b62ae6622dbbbe49ab4ffdc609f85c80/nekohtml-1.9.6.2.jar:/Users/tlong/.gradle/caches/transforms-1/files-1.1/asynclayoutinflater-28.0.0.aar/d21dd4918535ed3d8abe1ce9f57922d0/jars/classes.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/nekohtml/xercesMinimal/1.9.6.2/d1c5e063683a0e6f77cd5f051a9d4af48346fa6/xercesMinimal-1.9.6.2.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-stdlib-common/1.3.10/1b19d99229dcedad7caf50534dce38fe82845269/kotlin-stdlib-common-1.3.10.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/org.apache.maven/maven-model/2.2.1/c0a1c17436ec3ff5a56207c031d82277b4250a29/maven-model-2.2.1.jar:/Users/tlong/.gradle/caches/transforms-1/files-1.1/extensions-1.1.1.aar/6219a9bb377c61ff749bc613381202e8/jars/classes.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/android.arch.core/common/1.1.1/e55b70d1f5620db124b3e85a7f4bdc7bd48d9f95/common-1.1.1.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/org.apache.maven/maven-repository-metadata/2.2.1/98f0c07fcf1eeb213bef8d9316a9935184084b06/maven-repository-metadata-2.2.1.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/junit/junit/4.12/2973d150c0dc1fefe998f834810d68f278ea58ec/junit-4.12.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/org.hamcrest/hamcrest-core/1.3/42a25dc3219429f0e5d060061f71acb49bf010a0/hamcrest-core-1.3.jar:/Users/tlong/.gradle/caches/transforms-1/files-1.1/print-28.0.0.aar/a0f06bf5c5d94d3c72e5242fa9882597/jars/classes.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/androidx.annotation/annotation/1.0.0/45599f2cd5965ac05a1488fa2a5c0cdd7c499ead/annotation-1.0.0.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/com.squareup.okhttp3/okhttp/3.10.0/7ef0f1d95bf4c0b3ba30bbae25e0e562b05cf75e/okhttp-3.10.0.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/com.google.android.apps.common.testing.accessibility.framework/accessibility-test-framework/2.1/c2d6267c3ad0435ddf9e27a94a0e0af4ca3fdfbb/accessibility-test-framework-2.1.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/org.ow2.asm/asm/7.0/d74d4ba0dee443f68fb2dcb7fcdb945a2cd89912/asm-7.0.jar:/Users/tlong/.gradle/caches/transforms-1/files-1.1/support-fragment-28.0.0.aar/18c92ec3f926a7b8df7fb1e038408020/jars/classes.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/org.hamcrest/hamcrest-library/1.3/4785a3c21320980282f9f33d0d1264a69040538f/hamcrest-library-1.3.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/android.arch.lifecycle/common/1.1.1/207a6efae6a3555e326de41f76bdadd9a239cbce/common-1.1.1.jar:/Users/tlong/.gradle/caches/transforms-1/files-1.1/livedata-core-1.1.1.aar/7f00e7e1815ad61a0905cc6c72526e4d/jars/classes.jar:/Users/tlong/.gradle/caches/transforms-1/files-1.1/support-compat-28.0.0.aar/ae433b75533c5fdf809e409a362ea7e1/res:/Users/tlong/.gradle/caches/transforms-1/files-1.1/support-compat-28.0.0.aar/ae433b75533c5fdf809e409a362ea7e1/jars/classes.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/org.codehaus.plexus/plexus-interpolation/1.11/ad9dddff6043194904ad1d2c00ff1d003c3915f7/plexus-interpolation-1.11.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/org.codehaus.plexus/plexus-utils/1.5.15/c689598ce1eb94c304817877ed15911099972526/plexus-utils-1.5.15.jar:/Users/tlong/.gradle/caches/transforms-1/files-1.1/animated-vector-drawable-28.0.0.aar/8bb7b9cdff9c109d4ba17dbaa8969298/jars/classes.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/org.apache.maven/maven-artifact/2.2.1/23600f790d4dab2cb965419eaa982e3e84c428f8/maven-artifact-2.2.1.jar:/Users/tlong/.gradle/caches/transforms-1/files-1.1/monitor-1.1.0.aar/b0e7da0824083df03a4c0e0f0fcb38ef/jars/classes.jar:/Users/tlong/.gradle/caches/transforms-1/files-1.1/appcompat-v7-28.0.0.aar/d90e0247bae064c094288c2f78655953/res:/Users/tlong/.gradle/caches/transforms-1/files-1.1/appcompat-v7-28.0.0.aar/d90e0247bae064c094288c2f78655953/jars/classes.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/org.robolectric/shadows-framework/4.0.2/3bf4e4e4c8379dde73e95a15577ea48c799d9ac8/shadows-framework-4.0.2.jar:/Users/tlong/.gradle/caches/modules-2/files-2.1/classworlds/classworlds/1.1-alpha-2/5adf2e681c57d7f48038b602f3ca2254ee82d47/classworlds-1.1-alpha-2.jar:/Users/tlong/Documents/android-components/components/support/base/build/intermediates/javac/debug/compileDebugJavaWithJavac/classes:/Users/tlong/Documents/android-components/components/support/base/build/tmp/kotlin-classes/debug:/Users/tlong/Documents/android-components/components/support/base/build/generated/res/rs/debug:/Users/tlong/Documents/android-components/components/support/base/build/generated/res/resValues/debug:/Users/tlong/Documents/android-components/components/support/test/build/intermediates/javac/debug/compileDebugJavaWithJavac/classes:/Users/tlong/Documents/android-components/components/support/test/build/tmp/kotlin-classes/debug:/Users/tlong/Documents/android-components/components/support/test/build/generated/res/rs/debug:/Users/tlong/Documents/android-components/components/support/test/build/generated/res/resValues/debug:/Users/tlong/Documents/android-components/components/support/ktx/build/intermediates/javac/debug/compileDebugJavaWithJavac/classes:/Users/tlong/Documents/android-components/components/support/ktx/build/tmp/kotlin-classes/debug:/Users/tlong/Documents/android-components/components/support/ktx/build/generated/res/rs/debug:/Users/tlong/Documents/android-components/components/support/ktx/build/generated/res/resValues/debug:/Users/tlong/Documents/android-components/components/support/ktx/build/intermediates/sourceFolderJavaResources/debug:/Users/tlong/Documents/android-components/components/support/base/build/intermediates/sourceFolderJavaResources/debug:/Users/tlong/Documents/android-components/components/service/glean/build/intermediates/sourceFolderJavaResources/debug:/Users/tlong/Documents/android-components/components/support/test/build/intermediates/sourceFolderJavaResources/debug:/Users/tlong/.gradle/caches/transforms-1/files-1.1/android.jar/9e78d062dee754fae3593153726151f5/android.jar" com.intellij.rt.execution.application.AppMainV2 com.intellij.rt.execution.junit.JUnitStarter -ideVersion5 "mozilla.components.service.glean.storages.StringListsStorageEngineTest,string list deserializer should correctly parse string lists"
[Robolectric] mozilla.components.service.glean.storages.StringListsStorageEngineTest.string list deserializer should correctly parse string lists: sdk=28; resources=legacy
[Robolectric] NOTICE: legacy resources mode is deprecated; see http://robolectric.org/migrating/#migrating-to-40

org.json.JSONException: Value invalid_string of type java.lang.String cannot be converted to JSONArray

	at org.json.JSON.typeMismatch(JSON.java:112)
	at org.json.JSONArray.<init>(JSONArray.java:96)
	at org.json.JSONArray.<init>(JSONArray.java:108)
	at mozilla.components.service.glean.storages.StringListsStorageEngineImplementation.deserializeSingleMetric(StringListsStorageEngine.kt:44)
	at mozilla.components.service.glean.storages.StringListsStorageEngineImplementation.deserializeSingleMetric(StringListsStorageEngine.kt:26)
	at mozilla.components.service.glean.storages.GenericScalarStorageEngine.deserializeUserLifetime(GenericScalarStorageEngine.kt:126)
	at mozilla.components.service.glean.storages.GenericScalarStorageEngine$userLifetimeStorage$2.invoke(GenericScalarStorageEngine.kt:46)
	at mozilla.components.service.glean.storages.GenericScalarStorageEngine$userLifetimeStorage$2.invoke(GenericScalarStorageEngine.kt:39)
	at kotlin.SynchronizedLazyImpl.getValue(LazyJVM.kt:74)
	at mozilla.components.service.glean.storages.GenericScalarStorageEngine.getUserLifetimeStorage(GenericScalarStorageEngine.kt)
	at mozilla.components.service.glean.storages.GenericScalarStorageEngine.getSnapshot(GenericScalarStorageEngine.kt:156)
	at mozilla.components.service.glean.storages.StringListsStorageEngineTest.string list deserializer should correctly parse string lists(StringListsStorageEngineTest.kt:225)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.robolectric.RobolectricTestRunner$HelperTestRunner$1.evaluate(RobolectricTestRunner.java:600)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.robolectric.internal.SandboxTestRunner$2.evaluate(SandboxTestRunner.java:260)
	at org.robolectric.internal.SandboxTestRunner.runChild(SandboxTestRunner.java:130)
	at org.robolectric.internal.SandboxTestRunner.runChild(SandboxTestRunner.java:42)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.robolectric.internal.SandboxTestRunner$1.evaluate(SandboxTestRunner.java:84)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.intellij.rt.execution.application.AppMainV2.main(AppMainV2.java:131)


Process finished with exit code 255
```

Reply:
This exception happens when I do something like this, without the try/catch.  Specifically the line in the mapped file that has `toList("1", "2", "3")`:

```
    @Test
    fun `string list deserializer should correctly parse string lists`() {
        val persistedSample = mapOf(
                "store1#telemetry.invalid_string" to "invalid_string",
                "store1#telemetry.invalid_bool" to false,
                "store1#telemetry.null" to null,
                "store1#telemetry.invalid_int" to -1,
                "store1#telemetry.invalid_list" to listOf("1", "2", "3"),
                "store1#telemetry.invalid_int_list" to "[1,2,3]",
                "store1#telemetry.valid" to "[\"a\",\"b\",\"c\"]"
        )

        val storageEngine = StringListsStorageEngineImplementation()

        // Create a fake application context that will be used to load our data.
        val context = Mockito.mock(Context::class.java)
        val sharedPreferences = Mockito.mock(SharedPreferences::class.java)
        Mockito.`when`(sharedPreferences.all).thenAnswer { persistedSample }
        Mockito.`when`(context.getSharedPreferences(
                ArgumentMatchers.eq(storageEngine::class.java.simpleName),
                ArgumentMatchers.eq(Context.MODE_PRIVATE)
        )).thenReturn(sharedPreferences)

        storageEngine.applicationContext = context
        val snapshot = storageEngine.getSnapshot(storeName = "store1", clearStore = true)
        // Because JSONArray constructor will deserialize with or without the escaped quotes, it
        // treat the invalid_int_list above the same as the valid list, so we assertEquals 2
        assertEquals(2, snapshot!!.size)
        assertEquals(listOf("a", "b", "c"), snapshot["telemetry.valid"])
    }
```

Reply:
I see, got it, thanks! I think we should be safe with stuff from [this](https://github.com/mozilla-mobile/android-components/pull/1537#discussion_r241124219) comment then!